src/tools/tidy/src/deps.rs

```diff
@@ -46,30 +48,37 @@ type ExceptionList = &'static [(&'static str, &'static str)];
 /// * Optionally a tuple of:
 ///     * A list of crates for which dependencies need to be explicitly allowed.
 ///     * The list of allowed dependencies.
+/// * An indication of whether it is a submodule or not.
 // FIXME auto detect all cargo workspaces
-pub(crate) const WORKSPACES: &[(&str, ExceptionList, Option<(&[&str], &[&str])>)] = &[
+pub(crate) const WORKSPACES: &[(&str, ExceptionList, Option<(&[&str], &[&str])>, bool)] = &[
     // The root workspace has to be first for check_rustfix to work.
-    (".", EXCEPTIONS, Some((&["rustc-main"], PERMITTED_RUSTC_DEPENDENCIES))),
+    (".", EXCEPTIONS, Some((&["rustc-main"], PERMITTED_RUSTC_DEPENDENCIES)), false),
     // Outside of the alphabetical section because rustfmt formats it using multiple lines.
     (
         "compiler/rustc_codegen_cranelift",
         EXCEPTIONS_CRANELIFT,
         Some((&["rustc_codegen_cranelift"], PERMITTED_CRANELIFT_DEPENDENCIES)),
+        false,
     ),
     // tidy-alphabetical-start
-    ("compiler/rustc_codegen_gcc", EXCEPTIONS_GCC, None),
+    ("compiler/rustc_codegen_gcc", EXCEPTIONS_GCC, None, false),
     //("library/backtrace", &[], None), // FIXME uncomment once rust-lang/backtrace#562 has been synced back to the rust repo
     //("library/portable-simd", &[], None), // FIXME uncomment once rust-lang/portable-simd#363 has been synced back to the rust repo
     //("library/stdarch", EXCEPTIONS_STDARCH, None), // FIXME uncomment once rust-lang/stdarch#1462 has been synced back to the rust repo
-    ("src/bootstrap", EXCEPTIONS_BOOTSTRAP, None),
-    ("src/ci/docker/host-x86_64/test-various/uefi_qemu_test", EXCEPTIONS_UEFI_QEMU_TEST, None),
+    ("src/bootstrap", EXCEPTIONS_BOOTSTRAP, None, false),
+    (
+        "src/ci/docker/host-x86_64/test-various/uefi_qemu_test",
+        EXCEPTIONS_UEFI_QEMU_TEST,
+        None,
+        false,
+    ),
     //("src/etc/test-float-parse", &[], None), // FIXME uncomment once all deps are vendored
-    ("src/tools/cargo", EXCEPTIONS_CARGO, None),
+    ("src/tools/cargo", EXCEPTIONS_CARGO, None, false),
     //("src/tools/miri/test-cargo-miri", &[], None), // FIXME uncomment once all deps are vendored
     //("src/tools/miri/test_dependencies", &[], None), // FIXME uncomment once all deps are vendored
-    ("src/tools/rust-analyzer", EXCEPTIONS_RUST_ANALYZER, None),
-    ("src/tools/rustc-perf", EXCEPTIONS_RUSTC_PERF, None),
-    ("src/tools/x", &[], None),
+    ("src/tools/rust-analyzer", EXCEPTIONS_RUST_ANALYZER, None, false),
+    ("src/tools/rustc-perf", EXCEPTIONS_RUSTC_PERF, None, true),
```

Comment:
nit: It should be relatively to find out if the path is a submodule, so we could automate it. But this information doesn't change often, so having it specified manually is probably fine.

Reply:
In my opinion it's not worth it because it can slow down tidy (in my experience `git` can be quite slow with submodules depending on the host machine and target submodule). I intentionally made it this way.

Reply:
Bootstrap just reads `.gitmodules` and checks if any record from it starts with the given path, which should be fast (no git operations are performed). We could even extract that code from bootstrap and put it under `build_helper`.

But I agree that it's not worth to do it here.